### PR TITLE
Migrate Sell > Catalog > Feature listing

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Migrate Sell > Catalog > Feature list is now migrated to Symfony controller and prestashop Grid component. Even though Features page was used for testing horizontal migration approach, in the end it was decided that horizontal migration is not worth the time. Some actions are still missing (will be added gradually over multiple PR's) - delete, bulk delete, export, positioning.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | 
| Fixed ticket?     | https://github.com/PrestaShop/PrestaShop/issues/10540 
| Related PRs       | 
| Sponsor company   | Your company or customer's name goes here (if applicable).

#### BC break
Core/Domain/Feature/ValueObject/FeatureId constructor now throws FeatureConstraintException if 0 is passed

### CODE REVIEW

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Features/index.html.twig
Comment (Hlavtox) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Features/index.html.twig]: `<div class="row"><div class="col">` have no effect I think?

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php:61]: Can you please leave this old horizontal action code for now, we're probably gonna need some elements from it when we'll handle removing the `legacyLayout` It'd be easier if we know where to find the code instead of going back to the history.

You can rename it to `legacyHorizontalIndexAction` it doesn't need to have a related route, and you can add a comment that explains why we keep it for now (until legacyLayout is refactored) and that it should be cleaned after it's done

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php:130]: Yep it's a better fix than force-casting it later, so now you can clean `(int) $featureId` to `$featureId`

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php:178]: All those should stay as well for now at least (everything related to the horizontal page will be cleaned later)

File: src/PrestaShopBundle/Resources/config/services/core/util/helper_card.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/util/helper_card.yml:97]: Nice! Where did you find those values? Were they in an issue?
Or did you manually search for each equivalent on the website?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Features/index.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Features/index.html.twig]: @Hlavtox has made a huge work on uniformizing the BO pages wording, including these buttons in this PR https://github.com/PrestaShop/PrestaShop/pull/31322

I think we should follow his example and keep the consistency, so the buttons should be initialized in the controller instead of the twig file (it's also more consistent with the work we've been doing on simplifying the forms, less in twig more in PHP)

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

Just a few comments, the most important is about the horizontal action code, we should keep it as an example for now, it will be cleaned later

File: src/PrestaShopBundle/Resources/config/services/core/util/helper_card.yml
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/services/core/util/helper_card.yml:97]: manually :smile:

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php
Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/FeatureController.php:61]: hmm are you sure? We can always check previous commits how it worked, its not used anymore as (I think) it cannot implement the bridge interface anymore and I may get php stan complaining about unused methods :thinking:

**Comment by zuk3975** [comment]
@jolelievre phpStan was still failing (because of legacy classes usage required by horizontal cotroller, which I introduced to have a working example for now as you asked), so I just ignored all of those errors in code with `/* @phpstan-ignore-next-line */`, I think its the only option and should be good (considering it is just an example controller, which I marked as deprecated too make sure nobody uses it)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php
Comment (FabienPapet) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php:1]: Should'nt this controller be part of some documentation instead of adding it to the core ? :/

Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php:1]: IMO it should not exist at all, but @jolelievre insists to keep a working example temporary until we are sure which parts of horizontal migration can still be used and which can be removed.

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php:1]: Indeed I requested it, a lot of work has been done during horizontal migration to be independent from the `AdminLegacyLayoutController` which is used in all migrated pages, this is the only page that we have to use as an example and I think it'll be useful when we tackle the legacyLayout migration

So instead of going back into git history or copy the code later I prefer it to be still available somewhere, the legacy layout will ba hanled in v9.0 so once it's done this will be removed anyway

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/features.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/features.yml]:
```suggestion
    _legacy_link: AdminFeatures
    _legacy_feature_flag: feature
```

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

Just one comment on routing to avoid the page from automatically replacing the legacy one for now

As this is a partial feature maybe we can merge it as is and wait for the grid actions to be implemented no? (delete mostly I guess?)

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/features.yml
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/features.yml]: https://github.com/PrestaShop/PrestaShop/pull/31689/commits/2c00425fa7e620edec854885106b3138722b349a
related to #31823

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php
Comment (FabienPapet) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/HorizontalFeatureController.php:1]: can we comment it so we can be sure this has no impact in any form ? It would be weird to have some phpstan issues, deprecated code causing ci fails on this "unused controller"

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975 

Since this is a partial PR this one can be merged once it gets two approvals, this page will be tested when the remaining missing actions are all integrated in the list

**Comment by jolelievre** [comment]
Running campaign just in case https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/4466819150

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

**Review by matks — APPROVED** [review]

**Comment by zuk3975** [comment]
automated tests
https://github.com/zuk3975/testing_pr/actions/runs/4678982368

File: src/Core/Domain/Feature/ValueObject/FeatureId.php
Comment (FabienPapet) [inline src/Core/Domain/Feature/ValueObject/FeatureId.php:66]:
```suggestion
        if (0 > $featureId) {
```

**Review by FabienPapet — DISMISSED** [review]

File: src/Core/Domain/Feature/ValueObject/FeatureId.php
Comment (zuk3975) [reply src/Core/Domain/Feature/ValueObject/FeatureId.php:66]: It was intentional. 0 is invalid for id.
And when used for a relation (to represent no relation case) we should use IdInterface approach with NoFeatureId class implementation (there are coulple examples of that already in some other domains)

**Comment by zuk3975** [comment]
ui tests campaing after sycncing tests branch https://github.com/zuk3975/testing_pr/actions/runs/4687653940

File: src/Core/Domain/Feature/ValueObject/FeatureId.php
Comment (zuk3975) [reply src/Core/Domain/Feature/ValueObject/FeatureId.php:66]: added to bc break list just now :heavy_check_mark:

**Comment by zuk3975** [comment]
@PrestaShop/qa-automation the UI scenarios still failing, but it doesn't seem to be related to the PR. Can you approve the failures are the same as in develop branch (and not from this PR) and can I merge this then? :thinking:

**Pull request merged**
